test/jdk/java/lang/String/concat/HiddenClassUnloading.java

```diff
@@ -0,0 +1,104 @@
+ * @run main/othervm -Xmx8M -Xms8M -Xverify:all -XX:-CompactStrings HiddenClassUnloading
+ */
+public class HiddenClassUnloading {
+    public static void main(String[] args) throws Throwable {
+        MethodHandles.Lookup lookup = MethodHandles.lookup();
+
+        Class<?>[] types = new Class[] {
+                int.class, long.class, double.class, float.class, char.class, boolean.class, String.class,
+        };
+        Object[] values = new Object[] {
+                1, 1L, 1D, 1F, 'C', true, "A",
+        };
+
+        for (int i = 0; i < 10_000; i++) {
+            int radix = types.length;
+            String str = Integer.toString(i, radix);
+            int length = str.length();
+            String recipe = "\1".repeat(length);
+            Class<?>[] ptypes = new Class[length];
+            Object[] pvalues = new Object[length];
+            for (int j = 0; j < length; j++) {
+                int index = Integer.parseInt(str.substring(j, j + 1), radix);
+                ptypes[j] = types[index];
+                pvalues[j] = values[index];
+            }
+            MethodType concatType = MethodType.methodType(String.class, ptypes);
+            CallSite callSite = StringConcatFactory.makeConcatWithConstants(
+                    lookup,
+                    "concat",
+                    concatType,
+                    recipe,
+                    new Object[0]
+            );
+            MethodHandle mh = callSite.dynamicInvoker();
+            String result = switch (length) {
+                case 1  -> (String) mh.invoke(pvalues[0]);
+                case 2  -> (String) mh.invoke(pvalues[0], pvalues[1]);
+                case 3  -> (String) mh.invoke(pvalues[0], pvalues[1], pvalues[2]);
+                case 4  -> (String) mh.invoke(pvalues[0], pvalues[1], pvalues[2], pvalues[3]);
+                case 5  -> (String) mh.invoke(pvalues[0], pvalues[1], pvalues[2], pvalues[3], pvalues[4]);
+                case 6  -> (String) mh.invoke(pvalues[0], pvalues[1], pvalues[2], pvalues[3], pvalues[4], pvalues[5]);
+                case 7  -> (String) mh.invoke(pvalues[0], pvalues[1], pvalues[2], pvalues[3], pvalues[4], pvalues[5], pvalues[6]);
+                case 8  -> (String) mh.invoke(pvalues[0], pvalues[1], pvalues[2], pvalues[3], pvalues[4], pvalues[5], pvalues[6], pvalues[7]);
+                case 9  -> (String) mh.invoke(pvalues[0], pvalues[1], pvalues[2], pvalues[3], pvalues[4], pvalues[5], pvalues[6], pvalues[7], pvalues[8]);
+                default -> throw new RuntimeException("length too large " + length);
+            };
+
+            StringBuilder sb = new StringBuilder();
+            for (int j = 0; j < pvalues.length; j++) {
+                sb.append(pvalues[j]);
+            }
+            assertEquals(sb.toString(), result);
+        }
+
+        long unloadedClassCount = ManagementFactory.getClassLoadingMXBean().getUnloadedClassCount();
+        if (unloadedClassCount == 0) {
```

Comment:
Sample `getUnloadedClassCount()` before going into the loop so that we check that there's progress.